x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java

```diff
@@ -133,9 +136,51 @@ private static Decision debugYesAllowed(RoutingAllocation allocation, List<Strin
      * Given a string of comma-separated prioritized tiers (highest priority
      * first) and an allocation, find the highest priority tier for which nodes
      * exist. If no nodes for any of the tiers are available, returns an empty
-     * {@code Optional<String>}.
+     * {@code Optional<String>}. This method takes into account the desired nodes
+     * in order to know if there are planned topology changes in the cluster
+     * that can remove a tier that's part of the cluster now.
      */
-    public static Optional<String> preferredAvailableTier(List<String> prioritizedTiers, DiscoveryNodes nodes) {
+    public static Optional<String> preferredAvailableTier(
+        List<String> prioritizedTiers,
+        DiscoveryNodes nodes,
+        DesiredNodes.MembershipInformation desiredNodes
+    ) {
+        final var desiredNodesPreferredTier = getPreferredTierFromDesiredNodes(prioritizedTiers, nodes, desiredNodes);
+
+        if (desiredNodesPreferredTier.isPresent()) {
+            return desiredNodesPreferredTier;
+        }
+
+        return getPreferredAvailableTierFromClusterMembers(prioritizedTiers, nodes);
+    }
+
+    /**
+     * Given a list of tiers in descending order, return the tier that's present
+     * in the desired nodes with the highest priority, if none is present returns an
+     * {@code Optional.empty()}.
+     */
+    public static Optional<String> getPreferredTierFromDesiredNodes(
```

Comment:
I wonder if we should wait for all seeing all expected nodes in higher prioritized tiers before we consider a tier removed? I worry that a user removes warm tier and upsizes hot tier in one orchestration change. And we immediately start pushing too much data to hot tier, making it run out of space before we see the new node(s)? Might not be super realistic (takes time to move data) but feels like a problem we should consider tackling.